File: common-core/src/test/java/io/quartic/common/client/WebsocketListenerShould.java
package io.quartic.common.client;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.quartic.common.SweetStyle;
import io.quartic.common.test.websocket.WebsocketServerRule;
import org.hamcrest.Matchers;
import org.immutables.value.Value;
import org.junit.Rule;
import org.junit.Test;
import rx.observers.TestSubscriber;

import static io.quartic.common.serdes.ObjectMappers.encode;
import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.hamcrest.Matchers.contains;
import static org.junit.Assert.assertThat;

public class WebsocketListenerShould {
    private static final int TIMEOUT_MILLISECONDS = 1000;

    @Rule
    public WebsocketServerRule server = new WebsocketServerRule();

    @Test
    public void emit_items_from_socket() throws Exception {
        server.setMessages(encode(TestThingImpl.of("foo")), encode(TestThingImpl.of("bar")));

        final WebsocketListener<TestThing> listener = createListener();

        TestSubscriber<TestThing> subscriber = TestSubscriber.create();
        listener.observable().subscribe(subscriber);
        subscriber.awaitValueCount(2, TIMEOUT_MILLISECONDS, MILLISECONDS);

        assertThat(subscriber.getOnNextEvents(), contains(TestThingImpl.of("foo"), TestThingImpl.of("bar")));
    }

    @Test
    public void skip_undecodable_items() throws Exception {
        server.setMessages("bad", encode(TestThingImpl.of("bar")));

        final WebsocketListener<TestThing> listener = createListener();

        TestSubscriber<TestThing> subscriber = TestSubscriber.create();
        listener.observable().subscribe(subscriber);
        subscriber.awaitValueCount(1, TIMEOUT_MILLISECONDS, MILLISECONDS);

        assertThat(subscriber.getOnNextEvents(), contains(TestThingImpl.of("bar")));
    }

    @Test
    public void only_create_one_connection_if_multiple_subscribers() throws Exception {
        final WebsocketListener<TestThing> listener = createListener();

        TestSubscriber<TestThing> subA = TestSubscriber.create();
        TestSubscriber<TestThing> subB = TestSubscriber.create();
        listener.observable().subscribe(subA);
        listener.observable().subscribe(subB);
        subA.awaitValueCount(1, TIMEOUT_MILLISECONDS, MILLISECONDS);
        subB.awaitValueCount(1, TIMEOUT_MILLISECONDS, MILLISECONDS);

        assertThat(server.numConnections(), Matchers.equalTo(1));
    }

    @Test
    public void not_connect_to_websocket_if_no_subscribers() throws Exception {
        createListener();

        assertThat(server.numConnections(), Matchers.equalTo(0));
    }

    private WebsocketListener<TestThing> createListener() {
        return WebsocketListener.Factory.of(server.uri(), new WebsocketClientSessionFactory(getClass()))
                .create(TestThing.class);
    }

    @SweetStyle
    @Value.Immutable
    @JsonSerialize(as = TestThingImpl.class)
    @JsonDeserialize(as = TestThingImpl.class)
    interface TestThing {
        String name();
    }
}

File: common-core/src/test/kotlin/io/quartic/common/websocket/WebsocketListenerShould.kt
package io.quartic.common.websocket

import io.quartic.common.serdes.encode
import io.quartic.common.test.websocket.WebsocketServerRule
import org.hamcrest.Matchers.contains
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Rule
import org.junit.Test
import rx.observers.TestSubscriber
import java.util.concurrent.TimeUnit.MILLISECONDS

class WebsocketListenerShould {

    @get:Rule
    val server = WebsocketServerRule()

    val listener by lazy {
        WebsocketListener.Factory(server.uri, WebsocketClientSessionFactory(javaClass)).create(TestThing::class.java)
    }

    @Test
    fun emit_items_from_socket() {
        server.messages = listOf(encode(TestThing("foo")), encode(TestThing("bar")))

        val subscriber = TestSubscriber.create<TestThing>()
        listener.observable.subscribe(subscriber)
        subscriber.awaitValueCount(2, TIMEOUT_MILLISECONDS, MILLISECONDS)

        assertThat(subscriber.onNextEvents, contains(TestThing("foo"), TestThing("bar")))
    }

    @Test
    fun skip_undecodable_items() {
        server.messages = listOf("bad", encode(TestThing("bar")))

        val subscriber = TestSubscriber.create<TestThing>()
        listener.observable.subscribe(subscriber)
        subscriber.awaitValueCount(1, TIMEOUT_MILLISECONDS, MILLISECONDS)

        assertThat(subscriber.onNextEvents, contains(TestThing("bar")))
    }

    @Test
    fun only_create_one_connection_if_multiple_subscribers() {
        val subA = TestSubscriber.create<TestThing>()
        val subB = TestSubscriber.create<TestThing>()
        listener.observable.subscribe(subA)
        listener.observable.subscribe(subB)

        assertThat(server.numConnections, equalTo(1))
    }

    @Test
    fun not_connect_to_websocket_if_no_subscribers() {
        assertThat(server.numConnections, equalTo(0))
    }

    @Test
    fun close_socket_on_unsubscribe() {
        listener.observable.subscribe().unsubscribe()

        assertThat(server.numDisconnections, equalTo(1))
    }

    @Test
    fun not_close_socket_if_only_one_of_two_subscribers_unsubscribes() {
        listener.observable.subscribe()
        listener.observable.subscribe().unsubscribe()

        assertThat(server.numDisconnections, equalTo(0))
    }

    data class TestThing(val name: String)

    companion object {
        private val TIMEOUT_MILLISECONDS: Long = 1000
    }
}
